lib/Dialect/OM/Evaluator/Evaluator.cpp

```diff
@@ -36,10 +39,126 @@ circt::om::getEvaluatorValuesFromAttributes(MLIRContext *context,
   return values;
 }
 
-/// Instantiate an Object with its class name and actual parameters.
-FailureOr<std::shared_ptr<evaluator::ObjectValue>>
-circt::om::Evaluator::instantiate(
-    StringAttr className, ArrayRef<evaluator::EvaluatorValuePtr> actualParams) {
+LogicalResult circt::om::evaluator::EvaluatorValue::finalize() {
+  using namespace evaluator;
+  // Early return if already finalized.
+  if (finalized)
+    return success();
+  // Enable the flag to avoid infinite recursions.
+  finalized = true;
+  assert(isFullyEvaluated());
+  return llvm::TypeSwitch<EvaluatorValue *, LogicalResult>(this)
+      .Case<AttributeValue, ObjectValue, ListValue, MapValue, ReferenceValue,
+            TupleValue>([](auto v) { return v->finalizeImpl(); });
+}
+
+Type circt::om::evaluator::EvaluatorValue::getType() const {
+  Type actualParamType;
+  if (auto *attr = dyn_cast<evaluator::AttributeValue>(this)) {
+    if (auto typedActualParam = attr->getAttr().dyn_cast_or_null<TypedAttr>())
```

Comment:
is it possible for this to be null? also--I think we enforce at the ODS level that these are all TypedAttr, so maybe it could be cast_or_null?

Reply:
It seems non-nullness is guaranteed here so I changed to cast, thanks!